chore(cloudfront): small refactoring of the Origin API

skinny85 opened the pull request:
Change the Origin to an interface, from an abstract class, and change its `bind` protocol to return an `OriginBindConfig` interface.

This is in preparation for handling Origin Groups in #9109 - when time comes to handle Origin Groups, we will add a new (optional) property to `OriginBindConfig`, of type `CfnDistribution.OriginGroupProperty`, and handle it in `Distribution`.

BREAKING CHANGE: the property Origin.domainName has been removed

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: d4901c51c150131359b4708a260fdd3653cf3fcf
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b26a137c-5681-4def-99c2-44bf0757e271%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

njlynch commented on packages/@aws-cdk/aws-cloudfront/lib/private/cache-behavior.ts:
```diff
@@ -21,14 +20,10 @@ export interface CacheBehaviorProps extends BehaviorOptions {
-  constructor(private readonly props: CacheBehaviorProps) {
-    this.origin = props.origin;
+  constructor(originId: string, private readonly props: CacheBehaviorProps) {
+    this.originId = originId;
```
Is this change due to potential conflicts with the ordering of calls to bind()?

This _is_ an internal class, but the interface feels pretty odd now:

```ts
  const behavior = new CacheBehavior('origin_id', {
    origin,
    pathPattern: '*',
  });
```

Seems a bit odd to pass in an Origin but not use it, and pass in the Origin id.

Maybe this suggests that this should take in a (possibly renamed) `AddCacheBehaviorProps` (without the Origin). It's a nitpick, but it would clean up the tests a bit.

skinny85 replied:
Yep, you're absolutely right! Changed.

njlynch commented on packages/@aws-cdk/aws-cloudfront/test/private/cache-behavior.test.ts:
```diff
@@ -4,25 +4,23 @@ import { AllowedMethods, HttpOrigin } from '../../lib';
 test('renders the minimum template with an origin and path specified', () => {
   const origin = new HttpOrigin('www.example.com');
-  const behavior = new CacheBehavior({
+  const behavior = new CacheBehavior('origin_id', {
     origin,
```
As I said above, this feels a little weird. It's just for the sake of the tests, but worth considering... does this class really need the origin at this point?

skinny85 replied:
Done!

skinny85 commented on packages/@aws-cdk/aws-cloudfront-origins/lib/s3-origin.ts:
```diff
@@ -26,39 +26,22 @@ export interface S3OriginProps {
-    if (bucket.isWebsite) {
-      proxyOrigin = new cloudfront.HttpOrigin(bucket.bucketWebsiteDomainName, {
+    this.origin = bucket.isWebsite
+      ? new cloudfront.HttpOrigin(bucket.bucketWebsiteDomainName, {
```
Not me, I looooove my ternaries 🙂

skinny85 replied:
Having said that, I actually reverted that change 🤪. I had a change of heart, and decided minimizing the changes in this PR is more important than my personal preferences.

skinny85 commented on packages/@aws-cdk/aws-cloudfront/lib/origin.ts:
```diff
@@ -4,12 +4,42 @@ import { CfnDistribution } from './cloudfront.generated';
+ * @experimental
+ */
+export interface OriginBindOptions {
+  /** dsadas ds */
```
Yup! Done.

skinny85 commented on packages/@aws-cdk/aws-cloudfront/lib/origin.ts:
```diff
@@ -97,30 +110,20 @@ export abstract class Origin {
-    const s3OriginConfig = this.renderS3OriginConfig();
-    const customOriginConfig = this.renderCustomOriginConfig();
-
-    if (!s3OriginConfig && !customOriginConfig) {
-      throw new Error('Subclass must override and provide either s3OriginConfig or customOriginConfig');
-    }
```
I feel like this is not the correct condition as we expand what is returned in `OriginBindConfig`, but you're right, let's cross that bridge when we get there.

skinny85 replied:
Reverted.

skinny85 commented:
@njlynch this is ready for another round of reviews. I've incorporated all of your comments, and I've also reverted the renaming of `Origin` to `OriginBase` and `OriginProps` to `CommonOriginProps` - I figured minimizing the changes in this PR is more important. Let me know what you think!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: 934a50ced5ee13e0b2109e39662ebda4359b0e61
* Result: FAILED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2fc7a5cb-0d8d-4c8e-a95e-cef1cb57c5f6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject6AEA49D1-qxepHUsryhcu
* Commit ID: fbc016f1d47200e424589cc4b716dc283fec8927
* Result: SUCCEEDED
* [Build Logs](https://6sc75qyfl7.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=28e584b5-c4f8-45ea-acee-95a21ab98ea2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->
